ftplugin/forth.vim

```diff
@@ -0,0 +1,29 @@
+" Vim filetype plugin
+" Language:	Forth
+" Maintainer:	Johan Kotlinski <kotlinski@gmail.com>
+" Last Change:	2023 Apr 12
+" URL:		https://github.com/jkotlinski/forth.vim
+
+if exists("b:did_ftplugin")
+  finish
+endif
+let b:did_ftplugin = 1
+
+let s:cpo_save = &cpo
+set cpo&vim
+
+setlocal commentstring=\\\ %s
+setlocal comments=b:\\,s:.(,mb:\ ,e:)
+setlocal formatoptions-=t formatoptions+=croql
```

Comment:
this is just my lack of Vim knowledge... but how do I test that "comments" and "formatoptions" are working?
what behaviors do they change?
why are default formatoptions not ideal for Forth?

Reply:
That `comments` value should be for `(` not `.(`:

```vim
setlocal comments=b:\\,s:(,mb:\ ,e:)
```

I force-pushed a fix.

These lines are pretty much boiler plate for source code related ftplugins so you're probably already using them with other filetypes and just not familiar with how they're configured.  See: https://github.com/search?q=repo%3Avim%2Fvim%20croql&type=code 

`commentstring` is used as a template for adding fold markers.  So manually creating a fold with `:set foldmethod=marker` and the `zf` command should insert markers that look like `\ {{{` and `\ }}}` at the start and end of the fold.  Some third-party plugins also use this value for various features.

`formatoptions` is altered to disable automatic wrapping of source code and enable automatic formatting of comments using the `comments` setting.  See `:help fo-table` for details.

`comments` is also used by other features such as searching for definitions and identifiers while skipping comments. 

A quick test would be to open a Forth file, `set textwidth=20` then insert a line comment longer than that.  The text should auto-wrap with the `\` inserted at the start of the next line.

You can also run `o` or `O` in normal mode on an existing comment line and, likewise, the `\` should be inserted at the start of the newly opened line to continue the comment.

Reply:
Thank you for the explanation! It seems to work just like intended.

Would it be OK to leave `formatoptions` untouched? Basically, I feel like the default options are just fine for Forth. The `t` option is helpful, since the code may be exported to systems that only support a given fixed width.

Also, I feel like some of these behaviours ought to be a user preference, rather than in the filetype. Similar views appeared in [this old thread](https://groups.google.com/g/vim_dev/c/EKDS1PP4rPo?pli=1).

Reply:
Another thing... is this a bug? I set textwidth=10, but when entering a multi-line comment, I only get a line break after the first line. e.g.:

    ( 1 2 3 4
      5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 . . .

Reply:
> Would it be OK to leave `formatoptions` untouched? Basically, I feel like the default options are just fine for Forth. The `t` option is helpful, since the code may be exported to systems that only support a given fixed width.

Sure.  I assume you realise that this wrapping is unconditional so it'll split the line at invalid locations.  I usually handle this situation manually by adding a `colorcolumn` because the auto-wrapping drives me nuts but YMMV. :)

It seems `q` is in the default value.  I wonder if that's changed over the years.

> Also, I feel like some of these behaviours ought to be a user preference, rather than in the filetype. Similar views appeared in [this old thread](https://groups.google.com/g/vim_dev/c/EKDS1PP4rPo?pli=1).

It's an issue that comes up occasionally but not very often give the number of ftplugins that do the same.  People also complain when the changes haven't been set in the ftplugin like in [this recent issue](https://github.com/vim/vim/issues/10625).

It's hard to know how many users dislike some of these settings and have silently resolved the problem for themselves and how many users like and/or just expect them.

Reply:
> Another thing... is this a bug? I set textwidth=10, but when entering a multi-line comment, I only get a line break after the first line. e.g.:
> 
> ```
> ( 1 2 3 4
>   5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 . . .
> ```

Yes, formatting doesn't work well with a three-piece `comments` setting that uses a space for the middle part.  It has other uses though so we just use space and ignore the 'bug'.

I've been meaning to have a look at improving this feature...for many years.

Reply:
Well... about how to think about text width, I understand that people have different preferences and I think that by itself would be the strongest reason to keep it out of the ftplugin.

About Forth code in general: since there is no syntax, it generally does not matter at all if one word moves from one line to the next. Like, ideal Forth code can look like this:

![bild](https://user-images.githubusercontent.com/3259963/232338003-87eec4d8-a7d9-4650-90c8-8b08296268cd.png)